extlib/millhone/src/cmd/analyze_container.rs

```diff
@@ -185,3 +185,54 @@ fn buffer(mut reader: impl Read) -> Result<Vec<u8>> {
     reader.read_to_end(&mut buf).context("Read buffer")?;
     Ok(buf)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const MILLHONE_OUT: &str = r#"
+{
+  "discovered_jars": {
+    "blobs/sha256/61aed1a8baa251dee118b9ab203c1e420f0eda0a9b3f9322d67d235dd27a12ee": [
+      {
+        "kind": "v1.discover.binary.jar",
+        "path": "jackson-annotations-2.17.1.jar",
+        "fingerprints": {
+          "v1.raw.jar": "wjGJk8cvY4tpKcUC5r8YuO15Wfv+rVuyWANBYCUIeDs=",
+          "v1.class.jar": "t2Btr6rNrvzghM5Nud2uldRGVjw0/n5rK9j0xooQQyk=",
+          "sha_256": "/MrYLhMXLA5DhNtxV3IZybhjHAgg9LGNqqVwFvtmHHY=",
+          "v1.mavencentral.jar": "/KfvYZLJrQXQe8UNqZG/k3qErzo="
+        }
+      }
+    ],
+    "blobs/sha256/4d84019f77d1aa837a2cb4225bb79fc03a45ae5a247284dda07cfb9fb8077bd1": [
+      {
+        "kind": "v1.discover.binary.jar",
+        "path": "inner_directory/commons-email2-jakarta-2.0.0-M1.jar",
+        "fingerprints": {
+          "sha_256": "MuEcK3nOFuySTTg4HOJi3vvTpI9bYspfMHa9AK2merQ=",
+          "v1.mavencentral.jar": "6bzpyKql6Q+UxKQgm14pcP4wHGo=",
+          "v1.class.jar": "2wRGbMGyGRwEXqNm53h1YK8OO879kvzDxazmJXiAcfI=",
+          "v1.raw.jar": "QA4SAurtJeo+lx1Vqve5uQYnvDKLFx1NgsSuoWKi8pw="
+        }
+      }
+    ],
+    "blobs/sha256/cc2447e1835a40530975ab80bb1f872fbab0f2a0faecf2ab16fbbb89b3589438": []
+  }
+}
+"#;
+
+    fn expected_jar_analysis() -> JarAnalysis {
+        serde_json::from_str(MILLHONE_OUT).expect("Read expected millhone output.")
+    }
+
+    #[test]
+    fn it_finds_expected_output() {
+        let image_tar_file =
+            PathBuf::from("../../test/App/Fossa/Container/testdata/jar_test_container.tar");
+
+        let res = jars_in_container(&image_tar_file).expect("Read jars out of container image.");
+
+        assert_eq!(res, expected_jar_analysis());
```

Comment:
This is _fine_, but I **strongly** recommend not using the built-in types for this!

If you use the actual type in the test harness that the application uses, and the type changes its serialization, your test won't notice that you've broken backwards compatibility.

I strongly recommend instead testing against the raw JSON value:
```rust
let expected = serde_json::json!({
  "discovered_jars": {
    "blobs/sha256/61aed1a8baa251dee118b9ab203c1e420f0eda0a9b3f9322d67d235dd27a12ee": [
      {
        "kind": "v1.discover.binary.jar",
        "path": "jackson-annotations-2.17.1.jar",
        "fingerprints": {
          "v1.raw.jar": "wjGJk8cvY4tpKcUC5r8YuO15Wfv+rVuyWANBYCUIeDs=",
          "v1.class.jar": "t2Btr6rNrvzghM5Nud2uldRGVjw0/n5rK9j0xooQQyk=",
          "sha_256": "/MrYLhMXLA5DhNtxV3IZybhjHAgg9LGNqqVwFvtmHHY=",
          "v1.mavencentral.jar": "/KfvYZLJrQXQe8UNqZG/k3qErzo="
        }
      }
    ],
    "blobs/sha256/4d84019f77d1aa837a2cb4225bb79fc03a45ae5a247284dda07cfb9fb8077bd1": [
      {
        "kind": "v1.discover.binary.jar",
        "path": "inner_directory/commons-email2-jakarta-2.0.0-M1.jar",
        "fingerprints": {
          "sha_256": "MuEcK3nOFuySTTg4HOJi3vvTpI9bYspfMHa9AK2merQ=",
          "v1.mavencentral.jar": "6bzpyKql6Q+UxKQgm14pcP4wHGo=",
          "v1.class.jar": "2wRGbMGyGRwEXqNm53h1YK8OO879kvzDxazmJXiAcfI=",
          "v1.raw.jar": "QA4SAurtJeo+lx1Vqve5uQYnvDKLFx1NgsSuoWKi8pw="
        }
      }
    ],
    "blobs/sha256/cc2447e1835a40530975ab80bb1f872fbab0f2a0faecf2ab16fbbb89b3589438": []
  }
});

let image_tar_file = PathBuf::from("../../test/App/Fossa/Container/testdata/jar_test_container.tar");
let res = jars_in_container(&image_tar_file).expect("Read jars out of container image");
let res = serde_json::to_value(res).expect("encode as json");

// Optional, but `pretty_assertions` makes for dramatically better test failures
pretty_assertions::assert_eq!(expected, res);
```

Reply:
Thanks, that was a good call out. I've changed the test to do this.